java/tables/src/main/java/sleeper/table/index/dynamodb/DynamoDBTableIndex.java

```diff
@@ -0,0 +1,129 @@
+import com.amazonaws.services.dynamodbv2.model.TransactWriteItemsRequest;
+import com.amazonaws.services.dynamodbv2.model.TransactWriteItemsResult;
+import com.amazonaws.services.dynamodbv2.model.TransactionCanceledException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import sleeper.configuration.properties.instance.InstanceProperties;
+import sleeper.core.table.TableAlreadyExistsException;
+import sleeper.core.table.TableId;
+import sleeper.core.table.TableIdGenerator;
+import sleeper.core.table.TableIndex;
+
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+import static sleeper.configuration.properties.instance.CdkDefinedInstanceProperty.TABLE_ID_INDEX_DYNAMO_TABLENAME;
+import static sleeper.configuration.properties.instance.CdkDefinedInstanceProperty.TABLE_NAME_INDEX_DYNAMO_TABLENAME;
+import static sleeper.dynamodb.tools.DynamoDBAttributes.createStringAttribute;
+import static sleeper.dynamodb.tools.DynamoDBUtils.streamPagedItems;
+
+public class DynamoDBTableIndex implements TableIndex {
+    private static final Logger LOGGER = LoggerFactory.getLogger(DynamoDBTableIndex.class);
+
+    public static final String TABLE_NAME_FIELD = DynamoDBTableIdFormat.TABLE_NAME_FIELD;
+    public static final String TABLE_ID_FIELD = DynamoDBTableIdFormat.TABLE_ID_FIELD;
+
+    private final AmazonDynamoDB dynamoDB;
+    private final String nameIndexDynamoTableName;
+    private final String idIndexDynamoTableName;
+    private final TableIdGenerator idGenerator = new TableIdGenerator();
+
+    public DynamoDBTableIndex(AmazonDynamoDB dynamoDB, InstanceProperties instanceProperties) {
+        this.dynamoDB = dynamoDB;
+        this.nameIndexDynamoTableName = instanceProperties.get(TABLE_NAME_INDEX_DYNAMO_TABLENAME);
+        this.idIndexDynamoTableName = instanceProperties.get(TABLE_ID_INDEX_DYNAMO_TABLENAME);
+    }
+
+    @Override
+    public TableId createTable(String tableName) throws TableAlreadyExistsException {
+        TableId id = TableId.idAndName(idGenerator.generateString(), tableName);
+        TransactWriteItemsRequest request = new TransactWriteItemsRequest()
+                .withReturnConsumedCapacity(ReturnConsumedCapacity.TOTAL)
+                .withTransactItems(
+                        new TransactWriteItem().withPut(new Put()
+                                .withTableName(nameIndexDynamoTableName)
+                                .withItem(DynamoDBTableIdFormat.getItem(id))
+                                .withConditionExpression("attribute_not_exists(#tablename)")
+                                .withExpressionAttributeNames(Map.of("#tablename", TABLE_NAME_FIELD))),
+                        new TransactWriteItem().withPut(new Put()
+                                .withTableName(idIndexDynamoTableName)
+                                .withItem(DynamoDBTableIdFormat.getItem(id))
```

Comment:
As `DynamoDBTableIdFormat.getItem(id)` is used twice, I think it would be better to calculate it once and re-use it.

Reply:
Done